assignments/assignment7/faulty-oops.md

```diff
@@ -0,0 +1,78 @@
+Mem abort info:
+  ESR = 0x96000045
+  EC = 0x25: DABT (current EL), IL = 32 bits
+  SET = 0, FnV = 0
+  EA = 0, S1PTW = 0
+  FSC = 0x05: level 1 translation fault
+Data abort info:
+  ISV = 0, ISS = 0x00000045
+  CM = 0, WnR = 1
+user pgtable: 4k pages, 39-bit VAs, pgdp=0000000042628000
+[0000000000000000] pgd=0000000000000000, p4d=0000000000000000, pud=0000000000000000
+Internal error: Oops: 96000045 [#1] SMP
+Modules linked in: hello(O) faulty(O) scull(O)
+CPU: 0 PID: 160 Comm: sh Tainted: G           O      5.15.18 #1
+Hardware name: linux,dummy-virt (DT)
+pstate: 80000005 (Nzcv daif -PAN -UAO -TCO -DIT -SSBS BTYPE=--)
+pc : faulty_write+0x14/0x20 [faulty]
+lr : vfs_write+0xa8/0x2b0
+sp : ffffffc008d23d80
+x29: ffffffc008d23d80 x28: ffffff8002692640 x27: 0000000000000000
+x26: 0000000000000000 x25: 0000000000000000 x24: 0000000000000000
+x23: 0000000040000000 x22: 000000000000000c x21: 000000557b452690
+x20: 000000557b452690 x19: ffffff800264c400 x18: 0000000000000000
+x17: 0000000000000000 x16: 0000000000000000 x15: 0000000000000000
+x14: 0000000000000000 x13: 0000000000000000 x12: 0000000000000000
+x11: 0000000000000000 x10: 0000000000000000 x9 : 0000000000000000
+x8 : 0000000000000000 x7 : 0000000000000000 x6 : 0000000000000000
+x5 : 0000000000000001 x4 : ffffffc0006f7000 x3 : ffffffc008d23df0
+x2 : 000000000000000c x1 : 0000000000000000 x0 : 0000000000000000
+Call trace:
+ faulty_write+0x14/0x20 [faulty]
+ ksys_write+0x68/0x100
+ __arm64_sys_write+0x20/0x30
+ invoke_syscall+0x54/0x130
+ el0_svc_common.constprop.0+0x44/0xf0
+ do_el0_svc+0x40/0xa0
+ el0_svc+0x20/0x60
+ el0t_64_sync_handler+0xe8/0xf0
+ el0t_64_sync+0x1a0/0x1a4
+Code: d2800001 d2800000 d503233f d50323bf (b900003f) 
+---[ end trace 4ece37f3b4594c8d ]---
+```
+
+# PC
+
+```
+pc : faulty_write+0x14/0x20 [faulty]
+```
+
+# OBJDMP LOGS
+
+```
+Disassembly of section .text:
+
+0000000000000000 <faulty_write>:
+   0:	d503245f 	bti	c
+   4:	d2800001 	mov	x1, #0x0                   	// #0
+   8:	d2800000 	mov	x0, #0x0                   	// #0
+   c:	d503233f 	paciasp
+  10:	d50323bf 	autiasp
+  14:	b900003f 	str	wzr, [x1]
+  18:	d65f03c0 	ret
+  1c:	d503201f 	nop
+```
+
+# DESCRITPION
+
+    * The kernel crash occurred during the execution of the faulty_write function.
+    * The code at virtual address 0000000000000000 is attempting to write to memory using a NULL pointer (address 0), leading to a data abort.
+    * The memory abort information indicates a data abort with Error Status Register (ESR) 0x96000045, which corresponds to a level 1 translation fault.
+    * The Oops message mentions an internal error with code 96000045 and identifies the faulty function as faulty_write.
+    * The disassembly of faulty_write reveals a straightforward function that doesn't appear to contain an obvious issue. However, it might be interacting with data structures or pointers that have become corrupted or uninitialized, causing the crash.
```

Comment:
See the relevant code at https://github.com/cu-ecen-aeld/assignment-7-Suhas-Reddy-S/blob/edbf7080d952c8b232c52e286519220865134147/misc-modules/faulty.c#L53 where it dereferences a null pointer, as well as the store address at 0x14 above which references a pointer value in x1 set to 0 at offset 0x04.

The 0x14/0x20 indicates the null pointer occurred at the instruction at offset 0x14.